File: python/src/oset/ecg/peak_detection/peak_detection_simple.py
import argparse

import numpy as np


def peak_det_simple(x, ff, flag=0, omit_close_peaks=0):
    """
    peak_det_simple - Internal R-peak detector function
    Syntax: peaks, peak_indexes = peak_det_simple(x, ff, flag, omit_close_peaks)

      Inputs:
          x: Vector of input data
          ff: Approximate ECG beat-rate in Hertz
          flag: Search for positive (flag=1) or negative (flag=0) peaks
          omit_close_peaks: omit close peaks after main peak detection
          (true/1) or not(false/0). Default is 0

      Outputs:
          peaks: Vector of R-peak impulse train
          peak_indexes: Vector of R-peak indexes

    Revision History:
        July 2023: Translated to Python from Matlab (peak_det_simple.m)

    Amulya Jain, 2023
    The Open-Source Electrophysiological Toolbox
    https://github.com/alphanumericslab/OSET
    """

    n = len(x)
    peaks = np.zeros(n)
    rng = int(np.floor(0.5 / ff))

    if flag:
        for j in range(1, n + 1):
            # Determine the index range for peak search
            if rng < j < n - rng:
                index = slice(j - rng - 1, j + rng)
            elif j > rng:
                index = slice((n - 2 * rng) - 1, n)
            else:
                index = slice(0, 2 * rng)

            if np.max(x[index]) == x[j - 1]:
                peaks[j - 1] = 1
    else:
        for j in range(1, n + 1):
            # Determine the index range for peak search
            if rng < j < n - rng:
                index = slice(j - rng - 1, j + rng)
            elif j > rng:
                index = slice(n - 2 * rng - 1, n)
            else:
                index = slice(0, 2 * rng)

            if np.min(x[index]) == x[j - 1]:
                peaks[j - 1] = 1

    if omit_close_peaks:
        I = np.where(peaks)[0]
        d = np.diff(I)
        peaks[(I[np.where(d < rng)[0]])] = 0
    peak_indexes = np.where(peaks)[0] + 1
    return peaks, peak_indexes


if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        description="""
    peak_det_simple - Internal R-peak detector function
    Syntax: peaks, peak_indexes = peak_det_simple(x, ff, flag, omit_close_peaks)

      Inputs:
          x: Vector of input data
          ff: Approximate ECG beat-rate in Hertz
          flag: Search for positive (flag=1) or negative (flag=0) peaks
          omit_close_peaks: omit close peaks after main peak detection
          (true/1) or not(false/0). Default is 0

      Outputs:
          peaks: Vector of R-peak impulse train
          peak_indexes: Vector of R-peak indexes

    Revision History:
        July 2023: Translated to Python from Matlab (peak_det_simple.m)

    """,
        formatter_class=argparse.RawTextHelpFormatter,
    )
    args = parser.parse_args()

File: matlab/tools/ecg/peak_detection_simple.m
function [peaks, peak_indexes] = peak_detection_simple(x, ff, flag, varargin)
%
% peak_detection_simple - A basic R-peak detector
%
%   Syntax: [peaks, peak_indexes] = peak_detection_simple(x, ff, flag, omit_close_peaks)
%
%   Inputs:
%       x: Vector of input data
%       ff: Approximate ECG beat-rate in Hertz
%       flag: Search for positive (flag=1) or negative (flag=0) peaks
%       omit_close_peaks: omit close peaks after main peak detection
%       (true/1) or not(false/0). Default is 0
%
%   Outputs:
%       peaks: Vector of R-peak impulse train
%       peak_indexes: Vector of R-peak indexes
% 
%   Revision History:
%       2006: First release
%       2023: Extracted from deprecated version PeakDetection()
%
%   Reza Sameni, 2006-2023
%   The Open-Source Electrophysiological Toolbox
%   https://github.com/alphanumericslab/OSET

% Set default values for optional arguments
if nargin > 3 && ~isempty(varargin{1})
    omit_close_peaks = varargin{1};
else
    omit_close_peaks = 0;
end

N = length(x);
peaks = zeros(1, N);

rng = floor(0.5 / ff);

if flag
    for j = 1 : N
        % Determine the index range for peak search
        if j > rng && j < N - rng
            index = j - rng : j + rng;
        elseif j > rng
            index = N - 2 * rng : N;
        else
            index = 1 : 2 * rng;
        end

        if max(x(index)) == x(j)
            peaks(j) = 1;
        end
    end
else
    for j = 1 : N
        % Determine the index range for peak search
        if j > rng && j < N - rng
            index = j - rng : j + rng;
        elseif j > rng
            index = N - 2 * rng : N;
        else
            index = 1 : 2 * rng;
        end

        if min(x(index)) == x(j)
            peaks(j) = 1;
        end
    end
end

% Remove fake peaks
if omit_close_peaks
    I = find(peaks);
    d = diff(I);
    peaks(I(d < rng)) = 0;
end
peak_indexes = find(peaks);
